File: reactor-java/src/main/java/com/github/mostroverkhov/r2/reactor/R2Client.java
package com.github.mostroverkhov.r2.reactor;

import com.github.mostroverkhov.r2.core.*;
import com.github.mostroverkhov.r2.core.internal.acceptor.ClientAcceptor;
import com.github.mostroverkhov.r2.core.internal.requester.ClientSetup;
import com.github.mostroverkhov.r2.core.internal.requester.SetupData;
import com.github.mostroverkhov.r2.reactor.internal.Assertions;
import com.github.mostroverkhov.r2.reactor.internal.RequesterBuilder;
import com.github.mostroverkhov.r2.reactor.monitoring.Monitored;
import com.github.mostroverkhov.r2.reactor.monitoring.MonitoringSupport;
import io.rsocket.RSocket;
import io.rsocket.transport.ClientTransport;
import io.rsocket.util.DefaultPayload;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

import java.util.function.Consumer;

import static io.rsocket.RSocketFactory.ClientRSocketFactory;

public class R2Client extends R2ClientFluentBuilder<
    ClientRSocketFactory,
    ClientAcceptorBuilder,
    ClientTransport,
    Mono<RequesterFactory>> implements Monitored<R2Client> {

  private final Assertions assertions = Assertions.forClient();
  private final InterceptorOptions interceptorOptions = new InterceptorOptions();
  private Function1<
      ? super ClientAcceptorBuilder,
      ? extends ClientAcceptorBuilder> configurer;
  private Metadata metadata;
  private ClientTransport clientTransport;
  private MonitoringSupport monitoringSupport;

  @NotNull
  @Override
  public R2Client metadata(@NotNull Metadata metadata) {
    this.metadata = metadata;
    return this;
  }

  @NotNull
  @Override
  public R2Client transport(ClientTransport clientTransport) {
    this.clientTransport = clientTransport;
    return this;
  }

  @NotNull
  @Override
  public R2Client configureAcceptor(
      @NotNull Function1<
          ? super ClientAcceptorBuilder,
          ? extends ClientAcceptorBuilder> f) {
    configurer = f;
    return this;
  }

  @NotNull
  public R2Client interceptors(
      @NotNull Consumer<InterceptorOptions> configurer) {
    configurer.accept(interceptorOptions);
    return this;
  }

  @Override
  public Mono<RequesterFactory> start() {
    assertState();

    @SuppressWarnings("ConstantConditions")
    ClientRSocketFactory rSocketFactory = withSetup(getRSocketFactory());
    ClientRSocketFactory monitoredRSocketFactory = withMonitoring(rSocketFactory);

    ClientAcceptorBuilder acceptorBuilder =
        new ClientAcceptorBuilder(
            interceptorOptions.requesters(),
            interceptorOptions.handlers());

    ClientAcceptorBuilder configuredBuilder =
        configurer
            .invoke(acceptorBuilder);
    ClientAcceptor<RSocket, RSocket> acceptor =
        configuredBuilder
            .build();

    DataCodec requesterCodec = configuredBuilder
        .codecs()
        .primary();

    Mono<RSocket> rSocket =
        monitoredRSocketFactory
            .acceptor(acceptor::accept)
            .transport(clientTransport)
            .start();

    Mono<RequesterFactory> requesterFactory =
        rSocket
            .map(rs -> new RequesterBuilder(rs, interceptorOptions.requesters()))
            .map(requesterBuilder -> requesterBuilder.codec(requesterCodec))
            .map(CoreRequesterBuilder::build);

    return requesterFactory;
  }

  @Override
  public R2Client monitor(MonitoringSupport monitoringSupport) {
    this.monitoringSupport = monitoringSupport;
    return this;
  }

  private ClientRSocketFactory withSetup(ClientRSocketFactory factory) {
    SetupData setup = ClientSetup.metaData(metadata);
    return factory
        .dataMimeType(setup.getDataType())
        .metadataMimeType(setup.getMetadataType())
        .setupPayload(
            DefaultPayload.create(
                setup.getData(),
                setup.getMetadata())
        );
  }

  private ClientRSocketFactory withMonitoring(ClientRSocketFactory factory) {
    if (monitoringSupport != null) {

      MonitoringSupport.RSocketMonitoring rSocketMonitoring =
          monitoringSupport.rSocket();
      factory.addClientPlugin(rSocketMonitoring.requester())
          .addServerPlugin(rSocketMonitoring.handler());

      MonitoringSupport.R2Monitoring r2Monitoring = monitoringSupport.r2();
      interceptors(opts ->
          opts.requester(r2Monitoring.requester())
              .handler(r2Monitoring.handler()));
    }
    return factory;
  }

  private void assertState() {
    assertions.assertArg(getRSocketFactory(), "RSocketFactory");
    assertions.assertArg(clientTransport, "Transport");
    assertions.assertArg(configurer, "AcceptorConfigurer");
    assertions.assertMonitoring(monitoringSupport);
  }
}

File: android/src/main/java/com/github/mostroverkhov/r2/android/R2Client.kt
package com.github.mostroverkhov.r2.android

import com.github.mostroverkhov.r2.core.R2ClientFluentBuilder
import com.github.mostroverkhov.r2.core.Metadata
import com.github.mostroverkhov.r2.core.internal.requester.metaData
import com.github.mostroverkhov.r2.core.RequesterBuilder
import com.github.mostroverkhov.r2.core.RequesterFactory
import io.reactivex.Single
import io.rsocket.android.RSocketFactory.ClientRSocketFactory
import io.rsocket.android.transport.ClientTransport
import io.rsocket.android.util.PayloadImpl

typealias FluentBuilder = R2ClientFluentBuilder<
        ClientRSocketFactory,
        AndroidClientAcceptorBuilder,
        ClientTransport,
        Single<RequesterFactory>>

typealias AcceptorConfigurer = (AndroidClientAcceptorBuilder) -> AndroidClientAcceptorBuilder

class R2Client : FluentBuilder() {
    private var configurer: AcceptorConfigurer? = null
    private var metadata: Metadata? = null
    private var clientTransport: ClientTransport? = null

    override fun metadata(metadata: Metadata): FluentBuilder {
        this.metadata = metadata
        return this
    }

    override fun transport(transport: ClientTransport): FluentBuilder {
        clientTransport = transport
        return this
    }

    override fun configureAcceptor(f: AcceptorConfigurer): FluentBuilder {
        configurer = f
        return this
    }

    override fun start(): Single<RequesterFactory> {
        assertState()
        val acceptorBuilder = AndroidClientAcceptorBuilder()
        val transport = clientTransport!!
        val configure = configurer!!

        val configuredAcceptorBuilder = configure(acceptorBuilder)
        val clientAcceptor = configuredAcceptorBuilder.build()
        val requesterCodec = configuredAcceptorBuilder.codecs().primary()

        val rSocket = withSetup(rSocketFactory)
                .acceptor { clientAcceptor::accept }
                .transport(transport)
                .start()

        return rSocket.map(::AndroidRequesterBuilder)
                .map { it.codec(requesterCodec) }
                .map(RequesterBuilder::build)
    }

    private fun assertState() {
        assertArg(clientTransport, "ClientRSocketFactory")
        assertArg(clientTransport, "ClientTransport")
        assertArg(configurer, "RequesterConfigurer")
    }

    private fun assertArg(arg: Any?, name: String) {
        if (arg == null) {
            throw IllegalArgumentException("$name was not set")
        }
    }

    private fun withSetup(factory: ClientRSocketFactory?): ClientRSocketFactory {
        val setupData = metaData(metadata)
        return factory
                ?.dataMimeType(setupData.dataType)
                ?.metadataMimeType(setupData.metadataType)
                ?.setupPayload(
                        PayloadImpl(
                                setupData.data,
                                setupData.metadata)
                )
                ?: throw IllegalArgumentException("ClientRSocketFactory not set")
    }
}